File: ios/Classes/AnchorInfo.swift
//
//  AnchorInfo.swift
//  RNAzureSpatialAnchors
//
//  Created by Pietro De Caro on 21/11/2019.
//  Copyright © 2019 Facebook. All rights reserved.
//

import Foundation

class Ticket {
    var id: String
    var title: String
    var ts: String
    var ARanchorID: String?
    
    init(val: [String: Any]) {
        id = val["id"] as! String
        title = val["title"] as! String
        ARanchorID = val["ar_anchor"] as? String
        ts = val["ts"] as! String
    }
}

class AnchorInfo: Hashable {
    var id: String
    //name: cod for Asset, title for Ticket
    var name: String
    //type: "asset" for Asset, "ticket" for Ticket
    var type: String
    var ts: String?
    var ARanchorID: String?
    var tickets: [AnchorInfo]?
  
    init(val: [String: Any]) {
        id = val["id"] as! String
        type = val["type"] as! String
        name = val["name"] as! String
        ARanchorID = val["ar_anchor"] as? String
        ts = val["ts"] as? String
        if let tkts = val["tickets"] as? [[String: Any]]{
            tickets = tkts.map { AnchorInfo(val: $0) }
        }
    }
  
    static func ==(lhs: AnchorInfo, rhs: AnchorInfo) -> Bool {
        return lhs.id == rhs.id
    }
  
    func hash(into hasher: inout Hasher) {
        hasher.combine(id)
    }
}

File: android/src/main/kotlin/io/carius/lars/ar_flutter_plugin/AnchorInfo.kt
package io.carius.lars.ar_flutter_plugin

import java.time.ZonedDateTime
import java.time.LocalDateTime
import kotlin.collections.ArrayList


internal class AnchorInfo(
    val id: String,
    //name: cod for Asset, title for Ticket
    val name: String,
    //type: "asset" for Asset, "ticket" for Ticket
    val type: String,
    val ts: String?,
    var ARanchorId: String?,
    var tickets: ArrayList<AnchorInfo>?
) {
    constructor(map: Map<String, Any>) : this(
        map["id"].toString(),
        map["name"].toString(),
        map["type"].toString(),
        if (map["ts"] != null) map["ts"].toString() else null,
        if (map["ar_anchor"] != null) map["ar_anchor"].toString() else null,
        if (map["tickets"] != null) ArrayList() else null
    ){
        if (map["tickets"] != null){
            val array = map["tickets"] as ArrayList<Map<String, Any>>;
            for (t in array){
                tickets!!.add(AnchorInfo(t))
            }
        }
    }

    @Override
    override fun toString(): String {
        return "AnchorInfo(id: $id, name: $name, type: $type, ar_anchor: ${ARanchorId}, tickets: ${tickets?.size.toString()})"
    }
}